eclair-core/src/main/scala/fr/acinq/eclair/wire/FailureMessage.scala

```diff
@@ -38,9 +41,22 @@ case object InvalidRealm extends Perm { def message = "realm was not understood
 case object TemporaryNodeFailure extends Node { def message = "general temporary failure of the processing node" }
 case object PermanentNodeFailure extends Perm with Node { def message = "general permanent failure of the processing node" }
 case object RequiredNodeFeatureMissing extends Perm with Node { def message = "processing node requires features that are missing from this onion" }
-case class InvalidOnionVersion(onionHash: ByteVector32) extends BadOnion with Perm { def message = "onion version was not understood by the processing node" }
-case class InvalidOnionHmac(onionHash: ByteVector32) extends BadOnion with Perm { def message = "onion HMAC was incorrect when it reached the processing node" }
-case class InvalidOnionKey(onionHash: ByteVector32) extends BadOnion with Perm { def message = "ephemeral key was unparsable by the processing node" }
+case class InvalidOnionVersion(onionHash: ByteVector32) extends BadOnion with Perm {
+  def failureCode = FailureMessageCodecs.BADONION | FailureMessageCodecs.PERM | 4
+  def message = "onion version was not understood by the processing node"
+}
+case class InvalidOnionHmac(onionHash: ByteVector32) extends BadOnion with Perm {
+  def failureCode = FailureMessageCodecs.BADONION | FailureMessageCodecs.PERM | 5
+  def message = "onion HMAC was incorrect when it reached the processing node"
+}
+case class InvalidOnionKey(onionHash: ByteVector32) extends BadOnion with Perm {
+  def failureCode = FailureMessageCodecs.BADONION | FailureMessageCodecs.PERM | 6
+  def message = "ephemeral key was unparsable by the processing node"
+}
+case class InvalidOnionUnknown(onionHash: ByteVector32) extends BadOnion with Perm {
```

Comment:
I found it strange that Bolt 4 doesn't have an explicit error for that...so indeed to be compliant we fallback to the generic BadOnion error. Since the trait is already named `BadOnion` it causes issues when I rename this to `BadOnion`...but maybe simply `InvalidOnion`?

Reply:
Ok, but that's still not compliant right? There is no spec-defined message with that code *and* an enclosed `sha256_of_onion`.

Reply:
You're right, but in fact we're never sending any `BadOnion` failures in failure messages because when a `BadOnion` error occurs this should be an `update_fail_malformed_htlc` that's returned instead.
Unless I'm missing something it seems like the spec is confusing here because Bolt 2 separates `update_fail_malformed_htlc` from `update_fail_htlc` and Bolt 4 makes it look like there's no separation between those two.
Should we update the spec in Bolt 4 to remove all the `BadOnion` errors from the failure message and mention that all BadOnion errors should instead use only the `update_fail_malformed_htlc` that needs `sha_of_onion` and `failureCode`?